### PR TITLE
cpu/cortexm_common: use mpu stack guard if DEVELHELP is enabled

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

the stack guard only minimally reduces context switch performance, but is a tremendous help while developing.
So this PR enables it if DEVELHELP is enabled, for platforms that support it.

### CODE REVIEW

**Comment by kaspar030** [comment]
this should show a stack issue in the unitteests.

**Comment by kaspar030** [comment]
> this should show a stack issue in the unitteests.

actually not, because unittests don't enable DEVELHELP. ;/

**Comment by bergzand** [comment]
> this should show a stack issue in the unitteests.

And `mpu_stack_guard` is a module, not a feature :(

**Comment by miri64** [comment]
Why not also add `ssp` (see https://github.com/RIOT-OS/RIOT/pull/13175)?

**Comment by kaspar030** [comment]
> And `mpu_stack_guard` is a module, not a feature :(

true, will fix

**Comment by kaspar030** [comment]
> true, will fix

fixed by adding the cortexm_mpu -> mpu_stack_guard map to features_modules.mk.

File: makefiles/features_modules.inc.mk [hunk 1]
```diff
@@ -28,3 +28,8 @@ endif
 
 # select cortexm_svc pseudomodule if the corresponding feature is used
 USEMODULE += $(filter cortexm_svc, $(FEATURES_USED))
+
+# use mpu_stack_guard if the feature is used
+ifneq (,$(filter cortexm_mpu,$(FEATURES_USED)))
+  USEMODULE += mpu_stack_guard
+endif
```
Comment (bergzand) [inline makefiles/features_modules.inc.mk:35]: enabling in `mpu_stack_guard` when `cortexm_mpu` is included feels like the reverse order of things, but I guess this is a limitation of the current build system.

Comment (kaspar030) [reply makefiles/features_modules.inc.mk:35]: agreed... there should be optional dependencies, as in, "use mpu_stack_guard if it and all its deps are available".

**Comment by kaspar030** [comment]
> Why not also add `ssp` (see #13175)?

one at a time. :)

File: makefiles/features_modules.inc.mk (hunk 1 above)
Comment (bergzand) [reply makefiles/features_modules.inc.mk:35]: Ok, then let's not block on this build system issue

**Comment by kaspar030** [comment]
- rebased to resolve conflict. I guess tensorflow needs larger stacksize first before this can land.

**Comment by maribu** [comment]
> I guess tensorflow needs larger stacksize first before this can land.

Which proves your point that this is a useful tool and should be enabled by default :-)

**Comment by kaspar030** [comment]
I tried the tensorflow test locally, on samr21-xpro and nrf52dk, manually enabling mpu_stack_guard. No issues (didn't trigger the guard).
I think this is good to go!

**Comment by benpicco** [comment]
This broke the `saml10-xpro` :( 

> 43d6c4147a8034a1a115a061a1b8ca0705c86222 is the first bad commit
```
commit 43d6c4147a8034a1a115a061a1b8ca0705c86222
Author: Kaspar Schleiser <kaspar@schleiser.de>
Date:   Thu Jun 25 15:57:55 2020 +0200

    features_modules: cortexm_mpu -> mpu_stack_guard

 makefiles/features_modules.inc.mk | 5 +++++
 1 file changed, 5 insertions(+)
```

**Comment by bergzand** [comment]
There is a difference between the cortex-m0+ and the cortex-m4(and higher) MPU. The cortex-m0+ minimum MPU region size is 256 bytes, whereas the cortex-m4 MPU minimum region size is 32 bytes.

What could be happening here is that on the cortex-m0+, the region is automatically increased to 256B, covering a larger area than initially expected.

**Comment by benpicco** [comment]
The SAML10 is cortex-m23 though

**Comment by bergzand** [comment]
> The SAML10 is cortex-m23 though

Whoops. That changes things a bit.

The MPU on the cortex-m23 has different configuration bits compared to the older cortex-m(1|3|4|7). There is a guard here making it a dummy for the cortex-m23:
https://github.com/RIOT-OS/RIOT/blob/4d65bc8e0ab46dcc2eeb11e0deffb6af61f47871/cpu/cortexm_common/mpu.c#L63-L74

But there is no such guard in the `mpu_enable()` function:

https://github.com/RIOT-OS/RIOT/blob/4d65bc8e0ab46dcc2eeb11e0deffb6af61f47871/cpu/cortexm_common/mpu.c#L36-L37

I think the MPU gets enabled, but without any configured regions. I don't know what the default is, but I guess it doesn't allow anything by default.

**Pull request merged**